nodejs/src/lib/inference-session-impl.ts

```diff
@@ -0,0 +1,222 @@
+          }
+          fetches[name] = null;
+        }
+
+        if (typeof arg2 === 'object') {
+          options = arg2;
+        } else if (typeof arg2 !== 'undefined') {
+          throw new TypeError('\'options\' must be an object.');
+        }
+      } else {
+        // decide whether arg1 is fetches or options
+        // if any output name is present and its value is valid OnnxValue, we consider it fetches
+        let isFetches = false;
+        const arg1Keys = Object.getOwnPropertyNames(arg1);
+        for (const name of this.outputNames) {
+          if (arg1Keys.indexOf(name) !== -1) {
+            const v = arg1[name];
+            if (v === null || v instanceof Tensor) {
+              isFetches = true;
+              isFetchesEmpty = false;
+              fetches[name] = v;
+            }
+          }
+        }
+
+        if (isFetches) {
+          if (typeof arg2 === 'object') {
+            options = arg2;
+          } else if (typeof arg2 !== 'undefined') {
+            throw new TypeError('\'options\' must be an object.');
+          }
+        } else {
+          options = arg1 as RunOptions;
+        }
+      }
+    } else if (typeof arg1 !== 'undefined') {
+      throw new TypeError('Unexpected argument[1]: must be \'fetches\' or \'options\'.');
+    }
+
+    // check if all inputs are in feed
+    for (const name of this.inputNames) {
+      if (typeof feeds[name] === 'undefined') {
+        throw new Error(`input '${name}' is missing in 'feeds'.`);
+      }
+    }
+
+    // if no fetches is specified, we use the full output names list
+    if (isFetchesEmpty) {
+      for (const name of this.outputNames) {
+        fetches[name] = null;
+      }
+    }
+
+    // feeds, fetches and options are prepared
+
+    // promise start here
+    //
+    //
+    return new Promise((resolve, reject) => {
+      process.nextTick(() => {
+        try {
+          const returnValue: {[name: string]: OnnxValue} = {};
+          const results = this.#session.run(feeds, fetches, options);
+          for (const key in results) {
+            returnValue[key] = new Tensor(results[key].type, results[key].data, results[key].dims);
+          }
+          resolve(returnValue);
+        } catch (e) {
+          // reject if any error is thrown
+          reject(e);
+        }
+      });
+    });
+  }
+
+  inputNames: readonly string[];
+  outputNames: readonly string[];
+
+  #session: Binding.InferenceSession;
+}
+
+export const impl: InferenceSessionFactory = {
+  create: (arg0: string|ArrayBufferLike|Uint8Array, arg1?: SessionOptions|number, arg2?: number, arg3?: SessionOptions):
+      Promise<InferenceSession> => {
+        // either load from a file or buffer
+        let loadFromFilePath = false;
+        let filePath: string;
+        let buffer: ArrayBufferLike;
+        let byteOffset = -1;
+        let byteLength = -1;
+        let options: SessionOptions = {};
+
+        if (typeof arg0 === 'string') {
```

Comment:
May be more readable to get `typeof` and then use `switch`.

Reply:
because of JS's fantastic type system, we cannot simply use switch on `typeof arg0`:
 - check whether it's an object: `typeof arg0 === 'object'`
 - check whether it's a string: `typeof arg0 === 'string'`
 - check whether it's a number: `typeof arg0 === 'number'`

so far so good. however:
 - check whether it's an array: `Array.isArray(arg0)`
 - check whether it's safe integer: `Number.isSafeInteger(arg0)`
 - check whether it's a typed array: `arg0 instanceof Uint8Array`

If you take a look at lines below there are different checks applied so I cannot make this a simple switch here.